### PR TITLE
[string] Skip allocation in reserveCapacity if smol

### PR DESCRIPTION
**Pull request by milseman**
[benchmark] Add reserveCapacity String benchmark to builder.

Add StringBuilderSmallReservingCapacity variant to measure the excess
overhead caused by calls to reserve capacity. If a string is small,
and the capacity is small, this will demonstrate overhead if an
allocation happens.

---

[string] Skip allocation in reserveCapacity if smol

If the requested capacity is small enough to fit in our small string
representation, don't allocate a UTF-16 buffer, instead just return
early.

### CODE REVIEW

**Comment by milseman** [comment]
@swift-ci please test

**Comment by milseman** [comment]
@swift-ci please smoke benchmark

File: stdlib/public/core/StringGuts.swift [hunk 1]
```diff
@@ -966,18 +966,22 @@ extension _StringGuts {
       }
     }
 
-    // TODO (TODO: JIRA): check if we're small and still within capacity
+    // Small strings can accomodate small capacities
```
Comment (milseman) [inline stdlib/public/core/StringGuts.swift]: Note to self: use a spell checker. "accommodate"

**Comment by milseman** [comment]
@swift-ci please test linux platform

File: stdlib/public/core/StringGuts.swift (hunk 1 above)
Comment (natecook1000) [reply stdlib/public/core/StringGuts.swift]: It allocc

but it also accommoddatt

**Comment by milseman** [comment]
@swift-ci please test

**Pull request merged**